src/main/java/com/google/roomies/data/Listing.java

```diff
@@ -0,0 +1,201 @@
+import static com.google.roomies.ListingRequestParameterNames.NUM_SHARED;
+import static com.google.roomies.ListingRequestParameterNames.NUM_SINGLES;
+import static com.google.roomies.ListingRequestParameterNames.SHARED_ROOM_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.SINGLE_ROOM_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.LISTING_PRICE;
+import static com.google.roomies.ListingRequestParameterNames.START_DATE;
+import static com.google.roomies.ListingRequestParameterNames.TIMESTAMP;
+import static com.google.roomies.ListingRequestParameterNames.TITLE;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.Timestamp;
+import com.google.common.collect.ImmutableMap;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.io.Serializable;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import javax.money.CurrencyUnit;
+import javax.money.Monetary;
+import javax.servlet.http.HttpServletRequest;
+import org.javamoney.moneta.Money;
+
+/** A listing made by a user. */
+@AutoValue
+abstract class Listing implements Document, Serializable {
+  /**
+  * Describes the two different lease types for a listing.
+  * If input is not one of these two options, servlet throws an error.
+  */
+  enum LeaseType {
+    YEAR_LONG, 
+    MONTH_TO_MONTH;
+  }
+  abstract Optional<String> documentId();
+  abstract Optional<Timestamp> timestamp();
+  abstract String title();
+  abstract String description();
+  abstract Date startDate();
+  abstract Date endDate();
+  abstract LeaseType leaseType();
+  abstract int numRooms();
+  abstract int numBathrooms();
+  abstract int numShared();
+  abstract int numSingles();
+  abstract Money sharedPrice();
+  abstract Money singlePrice();
+  abstract Money listingPrice();
+
+  public static Builder builder() {
+    return new AutoValue_Listing.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder implements Serializable {
+    abstract Builder setTimestamp(Optional<Timestamp> timestamp);
+    abstract Builder setDocumentId(Optional<String> documentId);
+    abstract Builder setTitle(String title);
+    abstract Builder setDescription(String description);
+    abstract Builder setStartDate(Date startDate);
+    abstract Builder setEndDate(Date endDate);
+    abstract Builder setLeaseType(LeaseType leaseType);
+    abstract Builder setNumRooms(int numRooms);
+    abstract Builder setNumBathrooms(int numBathrooms);
+    abstract Builder setNumShared(int numShared);
+    abstract Builder setNumSingles(int numSingles);
+    abstract Builder setSharedPrice(Money sharedPrice);
+    abstract Builder setSinglePrice(Money singlePrice);
+    abstract Builder setListingPrice(Money listingPrice);
```

Comment:
I'm a little confused as to how these values work together - does a single posting have both singles and shared? Say a listing has 2 singles, one double, and one triple - how is that represented? Is that multiple postings? Or not allowed? Whatever answer is fine, but should be documented here.